proposals/accepted/047-standard-library-reform.rst

```diff
@@ -0,0 +1,559 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Based on the conversation in `CLC Issue #015`_, ``base`` is exposing too much stuff, yet trying to limit what is exposed would be a big breaking change.
+
+The solution is to reach for another layer of indirection.
+The CLC should be tasked with devising new standard library interfaces, which would initially be implemented by reexporting modules from ``base``.
+
+The new library interfaces should be carefully designed in and of themselves to tackle many, but not all, of the issues above:
+
+- They should be designed *not* to break every release.
+  Even though the underlying ``base`` from which modules are reexported would continue to have its regular problematic major version bumps, the portion reexported should have very infrequent breaking changes.
+
+  This fixes **Problem 1**.
+
+- These libraries should be emphasized in all documentation, and users should be encouraged to use them and not ``base`` in new end-application code.
+  ``base``, in contrast, would be kept exposed as a mere legacy interface.
+  As code migrates over to use the new standard libraries, ``base`` should become less important.
+  GHC devs can therefore feel increasingly confident modifying parts of ``base`` which are *not* reexported in these new libraries.
+
+  This partially fixes **Problem 2**.
+
+- The new standard library should not be a single library but multiple.
+  IO-free interfaces that are portable everywhere should be one library.
+  Interfaces involving IO should be split into libraries where they run.
+
+  For example, Unix and Windows are mostly a superset of WASI, so WASI-compatible file-descriptor-oriented code should work everywhere.
+
+  Exactly how many separate libraries is justified is left to the CLC to decide.
+
+  This fixes **Problem 3**.
+
+- Because these are new libraries "on top" of ``base``, they can also reexport items from libraries, like ``text``.
+  The CLC should consider such reexports.
+
+  This fixes **Problem 5**.
+
+New Goal: Rationalize dependencies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+**Step 1A** addresses most problems, but leaves behind **Problem 2** somewhat, and **Problem 4** completely.
+But moreover, **Step 1A** doesn't exactly make for a maintainable solution.
+As the famous David Wheeler quote states:
+"All problems in computer science can be solved by another level of indirection, *except for the problem of too many layers of indirection*."
+Reexporting modules from a less stable library (``base``) in more stable libraries is very error-prone.
+
+The generalization of these concerns is *rationalizing* dependencies, or rationalizing the division of labor between libraries.
+Once the purposes of libraries and the division of labor between them make more sense, it will be easier to maintain these libraries.
+It should be in fact easier than it was before to maintain them.
+
+New Goal: Split Base
```

Comment:
This section is quite uninformative to me.

Based on the rest of this document, the phrase "split base" could mean either of the following:
1. Cut base up with respect to functionality, so that platform-specific `IO` actions are in platform-specific `base` packages like `base-posix`, `base-windows`, `base-browser`, or `base-skynet`
2. Cut up base with respect to stability, so that the stable, reliable API is in `base` and the various GHC knobs and levers that are provided for advanced users but may change at any moment are in something like `ghc-base`

Which of these is meant here?

And why is this presented as a goal in and of itself, rather than a means to achieve one of the other goals?